### PR TITLE
Sous Chefs Adoption

### PR DESCRIPTION
**Pull request by TheREK3R**
# Description

Describe what this change achieves

## Issues Resolved

List any existing issues this PR resolves

## Check List

- [ ] All tests pass. See TESTING.md for details.
- [ ] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -23,76 +24,99 @@ jobs:
...
-          - 'ubuntu-1604'
-          - 'ubuntu-1804'
-          - 'ubuntu-2004'
+          - amazonlinux-2
+          - debian-9
+          - debian-10
+          - centos-7
+          - centos-8
+          - ubuntu-1804
+          - ubuntu-2004
```
Comment (ramereth) [inline .github/workflows/ci.yml:97]: This should only list the MacOS OS since we test the others via dokken.

File: .rubocop.yml [hunk 2]
```diff
@@ -1,2 +1,4 @@
----
-inherit_from: .rubocop_todo.yml
+AllCops:
+  TargetChefVersion: 16.latest
+Chef/Modernize/FoodcriticComments:
+  Enabled: true
```
Comment (ramereth) [inline .rubocop.yml:4]: Just remove this file.

File: Berksfile [hunk 3]
```diff
@@ -4,5 +4,5 @@ metadata
 
 # this group can be excluded by berks; for example: `berks upload --except test`
 group :test do
-  cookbook 'test', path: './test/fixtures/cookbooks/test'
+  cookbook 'splunk-test', path: './test/cookbooks/splunk-test'
```
Comment (ramereth) [inline Berksfile:7]: Let's not change this.

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,22 +2,32 @@
...
 - Fix centos-6 and Amazon linux convergence by adding an only_if to only create systemd symlinks on systemd systems.
 - Adds clarification text to README.md regarding chef-vault fallback to encrypted data bags
 - Changes the build status badge to track Actions status
 
 ## 7.0.1 (2020-10-30)
+
 - Moves most tests back to dokken and only run suites that change splunk user from root in vagrant.
 
 ## 7.0.0 (2020-10-22)
-**BREAKING CHANGE**
```
Comment (ramereth) [inline CHANGELOG.md:20]: Keep the `BREAKING CHANGE` change, but correct the formatting.

File: CHANGELOG.md [hunk 5]
```diff
@@ -26,18 +36,16 @@ This file is used to list changes made in each version of the splunk cookbook.
...
 - Fixes Test Kitchen configuration to test running Splunk as non-root user
 - Helper methods in `libraries/helper.rb` are moved to their own module space: `ChefSplunk::Helpers`
 - Disables Splunk management port (8089) when installing the Universal Forwarder
 
 ## 6.4.0 (2020-10-19)
+
 - Fixes Issue [#185](https://github.com/chef-cookbooks/chef-splunk/issues/185)
-  * a startup issue was resolved for SplunkForwarder installations with an improved
-    systemd unit file (fix below)
-  * Adds Inspec tests to verify from SplunkForwarder starts (thanks, [@jjm](https://github.com/jjm))
```
Comment (ramereth) [inline CHANGELOG.md:38]: Why did you remove the content above?

File: CHANGELOG.md [hunk 6]
```diff
@@ -46,199 +54,198 @@ This file is used to list changes made in each version of the splunk cookbook.
...
+
 - Drops travis-ci and onboards testing with Github Actions
 - removes dependency on the splunk command when running the `disable` recipe
 
 ## 6.2.8 (2020-07-01)
+
 - Thank you, [@doublethink](https://github.com/doublethink), for this submission
-- Resolves issues installing server and client as non-root users:
-  * `chef-splunk::user` recipe will not run if splunkd should be run as a non-root user
-  * systemd and SysV templates correctly run as the specified non-root user
```
Comment (ramereth) [inline CHANGELOG.md:71]: Sub-lists should be supported as long as they are formatted properly. Can you please fix this so that it still retains this but is correct according to `mdl`?

File: kitchen.dokken.yml [hunk 7]
```diff
@@ -1,58 +1,60 @@
...
-provisioner:
+transport:
   name: dokken
-  clean_dokken_sandbox: false
 
-transport:
+provisioner:
   name: dokken
+  deprecations_as_errors: true
+  chef_license: accept-no-persist
```
Comment (ramereth) [inline kitchen.dokken.yml:13]: These two should be already included since `kitchen.yml` also gets used.

File: kitchen.dokken.yml [hunk 8]
```diff
@@ -1,58 +1,60 @@
...
+
+verifier:
+  name: inspec
```
Comment (ramereth) [inline kitchen.dokken.yml:16]: Same with I stated above, this should be removed.

File: kitchen.dokken.yml [hunk 9]
```diff
@@ -1,58 +1,60 @@
...
+      intermediate_instructions:
+        - RUN /usr/bin/apt-get update
 
   - name: centos-7
     driver:
       image: dokken/centos-7
-      pid_one_command: /usr/lib/systemd/systemd
-      volumes:
-        - /sys/fs/cgroup:/sys/fs/cgroup:ro # required by systemd
+      pid_one_command: /sbin/init
```
Comment (ramereth) [inline kitchen.dokken.yml:41]:
```suggestion
      pid_one_command: /usr/lib/systemd/systemd
```

File: kitchen.yml [hunk 10]
```diff
@@ -1,9 +1,13 @@
 ---
+driver:
+  name: vagrant
+  boot_timeout: 600
```
Comment (ramereth) [inline kitchen.yml:4]: Is this really required?

File: kitchen.yml [hunk 11]
```diff
@@ -19,20 +23,27 @@ provisioner:
...
     attributes:
       splunk:
         accept_license: true
-    verifier:
-      name: inspec
-      inspec_tests:
-        - path: test/integration/inspec/client_test.rb
 
-  - name: client-runas-splunk
+  - name: client_runas_splunk
```
Comment (ramereth) [inline kitchen.yml:46]: Why change this from `-` to `_`?

File: metadata.rb [hunk 12]
```diff
@@ -1,21 +1,21 @@
...
+supports 'centos'
+supports 'debian'
+supports 'redhat'
+supports 'ubuntu'
 
 # please read the README.md section regarding data bag fallback if you
 # do not use chef-vault
+depends 'apt'
 depends 'chef-vault', '~> 4.0'
 depends 'ec2-tags-ohai-plugin', '~> 0.2.4'
```
Comment (ramereth) [inline metadata.rb:21]: I think this is a better way around this and frees up the dependency to folks.
```suggestion
depends 'chef-vault', '>= 4.0'
depends 'ec2-tags-ohai-plugin', '>= 0.2.4'
```

File: recipes/server.rb [hunk 13]
```diff
@@ -11,7 +11,7 @@
 #    http://www.apache.org/licenses/LICENSE-2.0
 #
 # Unless required by applicable law or agreed to in writing, software
-# distributed under the License is distributed on an "AS IS" BASIS,
+# distributed under the License is distributed on an "AS IS" BASIS
```
Comment (ramereth) [inline recipes/server.rb:14]: We shouldn't be modifying the license header like this, please revert.

**Review by ramereth — CHANGES_REQUESTED** [review]

File: recipes/server.rb (hunk 13 above)
Comment (TheREK3R) [reply recipes/server.rb:14]: i think this must have been an accident or cookstyle, sorry

File: kitchen.yml (hunk 11 above)
Comment (TheREK3R) [reply kitchen.yml:46]: I was having trouble with the testing and got confused with the inspec/kitchen folder structure, I can revert that

File: CHANGELOG.md (hunk 4 above)
Comment (TheREK3R) [reply CHANGELOG.md:20]: What is the correct formatting here? should this be part of the list or another header?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (TheREK3R) [reply .github/workflows/ci.yml:97]: This actually needs to test all of these because the runas-splunk recipes aren't working on dokken

File: CHANGELOG.md (hunk 5 above)
Comment (TheREK3R) [reply CHANGELOG.md:38]: I thought sub-lists werent allowed, Ill revert that

File: CHANGELOG.md (hunk 4 above)
Comment (ramereth) [reply CHANGELOG.md:20]: There needs to be a new line between the header and the text.

File: CHANGELOG.md (hunk 5 above)
Comment (ramereth) [reply CHANGELOG.md:38]: I don't believe so, as long as they are formatted correctly. You can check the Markdown spec though.

File: kitchen.yml (hunk 11 above)
Comment (ramereth) [reply kitchen.yml:46]: As long as the folders have the correct name, then I'd prefer to keep the dashes here.

File: kitchen.dokken.yml [hunk 14]
```diff
@@ -1,58 +1,60 @@
...
+  name: inspec
 
 platforms:
-  - name: debian-10
+  - name: amazonlinux-2
     driver:
-      image: dokken/debian-10
-      pid_one_command: /bin/systemd
-      volumes:
-        - /sys/fs/cgroup:/sys/fs/cgroup:ro # required by systemd
```
Comment (ramereth) [inline kitchen.dokken.yml:22]: This should be kept as this is what allows the `runas` suites to run properly on dokken.

File: metadata.rb [hunk 15]
```diff
@@ -1,21 +1,21 @@
...
 
 # please read the README.md section regarding data bag fallback if you
 # do not use chef-vault
-depends 'chef-vault', '~> 4.0'
-depends 'ec2-tags-ohai-plugin', '~> 0.2.4'
-
-source_url 'https://github.com/chef-cookbooks/chef-splunk'
-issues_url 'https://github.com/chef-cookbooks/chef-splunk/issues'
-chef_version '>= 13.11'
+depends 'apt'
```
[the remark this answers is not in the document]
Comment (Stromweld) [reply metadata.rb]: removed apt dependency

File: Berksfile [hunk 16]
```diff
@@ -4,5 +4,5 @@ metadata
 
 # this group can be excluded by berks; for example: `berks upload --except test`
 group :test do
-  cookbook 'test', path: './test/fixtures/cookbooks/test'
+  cookbook 'test', path: './test/fixtures/test'
```
[the remark this answers is not in the document]
Comment (Stromweld) [reply Berksfile]: Switched to policy file negates this change.

File: recipes/setup_shclustering.rb [hunk 17]
```diff
@@ -111,7 +111,7 @@
 
 ruby_block 'captain elected' do
   block do
-    node.normal['splunk']['shclustering']['captain_elected'] = true
+    node.default['splunk']['shclustering']['captain_elected'] = true
```
Comment (haidangwa) [inline recipes/setup_shclustering.rb:114]: this `ruby_block` sets the `node.normal` precedence on purpose. A splunk search head captain bootstrap procedure should only occur one time and one time only. Therefore, setting `node.normal` here is appropriate, since that value needs to be persisted on all future chef runs once the execute resource has bootstrapped the captain. If this is reverted to `node.default`, future runs that call helper method `#ok_to_bootstrap_captain?` will always be false.

File: resources/server_install.rb [hunk 18]
```diff
@@ -0,0 +1,112 @@
+# To learn more about Custom Resources, see https://docs.chef.io/custom_resources.html
+#
+# Author:: Corey Hemminger
```
Comment (haidangwa) [inline resources/server_install.rb:3]: I do believe some credit should be given to me here, as a lot of this code looks copy/pasted from recipes that I developed in the original repo.

File: resources/service.rb [hunk 19]
```diff
@@ -0,0 +1,53 @@
+# To learn more about Custom Resources, see https://docs.chef.io/custom_resources.html
+#
+# Author:: Corey Hemminger
```
Comment (haidangwa) [inline resources/service.rb:3]: I do believe some credit should be given to me here, as a lot of this code looks copy/pasted from recipes that I developed in the original repo.

File: recipes/setup_shclustering.rb (hunk 17 above)
Comment (Stromweld) [reply recipes/setup_shclustering.rb:114]: cookstyle barks about using normal.
```
recipes/setup_shclustering.rb:114:5: R: Chef/Correctness/NodeNormal: Do not use node.normal. Replace with default/override/force_default/force_override attribute levels. (https://docs.chef.io/workstation/cookstyle/chef_correctness_nodenormal)
    node.normal['splunk']['shclustering']['captain_elected'] = true
```

Comment (haidangwa) [reply recipes/setup_shclustering.rb:114]: that is why there has been an exception in `.rubocop_todo.yml`:

File: .rubocop_todo.yml [hunk 20]
```diff
@@ -1,15 +0,0 @@
...
-# `rubocop --auto-gen-config`
-# on 2020-06-30 20:30:26 -0700 using RuboCop version 0.75.1.
-# The point is for the user to remove these configuration records
-# one by one as the offenses are removed from the code base.
-# Note that changes in the inspected code, or installation of new
-# versions of RuboCop, may require this file to be generated again.
-
-# Offense count: 1
----
-ChefCorrectness/NodeNormal:
```
Comment (haidangwa) [inline .rubocop_todo.yml:11]: This file needs to be added back so cookstyle doesn't warn about `node.normal` use in the recipes/setup_shcluster.rb. In a normal use case, `node.normal` isn't recommended; however, there is a need to preserve an attribute value between chef runs. This is to prevent a bootstrap captain in a search head cluster from being bootstrapped once a captain has already been bootstrapped.

File: recipes/setup_shclustering.rb (hunk 17 above)
Comment (Stromweld) [reply recipes/setup_shclustering.rb:114]: I can't think of another way of doing this so I just used an in line ignore for cookstyle.

File: resources/server_install.rb (hunk 18 above)
Comment (Stromweld) [reply resources/server_install.rb:3]: agreed, that was copy pasted from another cookbook I had to start my resource file.

File: .rubocop_todo.yml (hunk 20 above)
Comment (Stromweld) [reply .rubocop_todo.yml:11]: I setup an inline ignore to fix the cookstyle error.

File: recipes/setup_shclustering.rb (hunk 17 above)
Comment (haidangwa) [reply recipes/setup_shclustering.rb:114]: what is wrong with using what was in the repo before -- `.rubocop.yml` and `.rubocop_todo.yml`?

Comment (haidangwa) [reply recipes/setup_shclustering.rb]: IMHO, inline ignores are code-smells. Hunting them down in the future is a maintenance nightmare.

Comment (Stromweld) [reply recipes/setup_shclustering.rb]: My thought is that then your ignoring the whole file so other places where it may not be absolutely needed or incorrectly used would also be ignored instead of the 1 specific edge case where it is needed.

Comment (Stromweld) [reply recipes/setup_shclustering.rb:114]: reverted my changes and moved my changes to a new fork https://github.com/TheREK3R/chef-splunk/pull/1

File: resources/server_install.rb (hunk 18 above)
Comment (Stromweld) [reply resources/server_install.rb]: reverted my changes and moved my changes to a new fork TheREK3R#1

File: .rubocop_todo.yml (hunk 20 above)
Comment (Stromweld) [reply .rubocop_todo.yml:11]: reverted my changes and moved my changes to a new fork TheREK3R#1

File: Berksfile (hunk 16 above)
[the remark this answers is not in the document]
Comment (TheREK3R) [reply Berksfile:7]: I forgot that was fixtures/cookbooks not just fixtures when moving it back

File: metadata.rb (hunk 15 above)
[the remark this answers is not in the document]
Comment (TheREK3R) [reply metadata.rb:19]: Dokken was throwing errors without it, but I'll go back and take a look again.

File: kitchen.yml [hunk 21]
```diff
@@ -1,9 +1,9 @@
 ---
 provisioner:
   deprecations_as_errors: true
+  multiple_converge: 2
```
Comment (ramereth) [inline kitchen.yml:4]: This isn't very useful without `enforce_idempotency: true`. Now I"m not sure how good this cookbook is with that. If it's a lot of work to fix idempotency, then I'd say skip it for this PR. But I'd try it locally on a few suites and see how far the rabbit hole is and if it's too big, then skip it and remove this line.

File: kitchen.yml [hunk 22]
```diff
@@ -19,6 +19,17 @@ provisioner:
...
+verifier: inspec
+
+platforms:
+  - name: amazonlinux-2
+  - name: debian-9
+  - name: debian-10
+  - name: centos-7
+  - name: centos-8
+  - name: ubuntu-18.04
+  - name: ubuntu-20.04
```
Comment (ramereth) [inline kitchen.yml:31]: Please alphabetize this.

File: test/integration/server-runas-root/inspec/server_runas_root_spec.rb [hunk 23]
```diff
@@ -5,6 +5,12 @@
   title 'Verify Enterprise Splunk server installation'
   only_if { os.linux? }
 
+  describe 'chef-splunk::server should run as "root" user' do
+    describe command('ps aux | grep "splunkd --under-systemd" | head -1 | awk \'{print $1}\'') do
```
Comment (ramereth) [inline test/integration/server-runas-root/inspec/server_runas_root_spec.rb:9]: Instead of using `ps` to check on processes, please use the [processes](https://docs.chef.io/inspec/resources/processes/) resource in Inspec. Please fix this globally as there are several places this has been done.

**Review by ramereth — CHANGES_REQUESTED** [review]

File: kitchen.yml (hunk 22 above)
Comment (TheREK3R) [reply kitchen.yml:31]: cant believe i didnt 🤦

File: kitchen.yml (hunk 21 above)
Comment (TheREK3R) [reply kitchen.yml]: I checked it and youre right, a lot fail with idempotency checks so I removed the multiple_converge

Comment (TheREK3R) [reply kitchen.yml]: This actually is nesecary for some tests that check on things that should exist after splunk is restarted. Should I leave in multiple converge or try to manually restart it during tests?

**Comment by haidangwa** [comment]
Yes, please leave it in. That was the main reason I added
multiple_converge, so that's would be able to pass street Splunk was
restarted. If you'd like to isolate that provisioner config to particular
test suites, that could cut down testing time.

File: kitchen.yml (hunk 21 above)
Comment (ramereth) [reply kitchen.yml:4]: Ok, let's leave it but leave a comment for why.

**Review by ramereth — APPROVED** [review]

**Comment by kitchen-porter** [comment]
Released as: [7.2.0](https://github.com/sous-chefs/chef-splunk/releases/tag/7.2.0)
